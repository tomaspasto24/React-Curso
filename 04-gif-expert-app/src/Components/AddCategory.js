import React from 'react'
import { useState } from 'react/cjs/react.development'
import PropTypes from 'prop-types';

export const AddCategory = ( { setCategories } ) => {

    //useState del valor que se encuentra en la barra de búsqueda.
    const [inputValue, setInputValue] = useState('');

    //Form llama a esta función cada vez que cambia el valor.
    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    //Form llama a esta función cuando se hace submit (por enter o por botón).
    const handleSubmit = (e) => {
        e.preventDefault();
        if (inputValue.trim().length > 2)
        {
            setCategories( (cats) => [inputValue, ...cats] );
            setInputValue('');
        }
    }

    return (
        <form onSubmit={ handleSubmit }>
            <p> {inputValue} </p>
          <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
          />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}
