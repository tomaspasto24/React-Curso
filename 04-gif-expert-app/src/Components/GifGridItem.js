import React from 'react';

export const GifGridItem = ({id, title, url}) => {

    console.log(id, title, url);

    return (
        //Se usa className porque class es una palabra reservada de js. Pero significan lo mismo.
        <div className="card animate__animated animate__fadeIn">
            <img src={url} alt={title}/>
            <p>{title}</p>
        </div>
    )
}