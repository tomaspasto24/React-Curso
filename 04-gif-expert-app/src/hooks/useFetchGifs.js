import { useEffect, useState } from "react";
import getGifs from "../helpers/getGifs";

const useFetchGifs = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    // Si category cambia se ejecuta el callback del 1er argumento.
    useEffect( () => {
        getGifs(category)
            .then( imgs => {
                setState({
                    data: imgs,
                    loading: false
                })
            })
    }, [category]);

    return state;
}

export default useFetchGifs;