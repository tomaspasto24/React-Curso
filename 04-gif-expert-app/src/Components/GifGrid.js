// import React, { useEffect } from 'react';
// import { useState } from 'react/cjs/react.development';
// import getGifts from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';
import React from 'react';
import useFetchGifs from '../hooks/useFetchGifs';


export const GifGrid = ( {category} ) => {

    const { data:images, loading } = useFetchGifs(category);
        
    return (
        <>
            <h3 className='card'> {category} </h3>

            {/* Evalua si loading es true */}
            { loading && <p>Loading...</p>}

            {/* Operador ternario: primera expersion si es true, segunda si es false. */}
            {/* {loading ? 'Cargando...' : 'Data cargada.'} */}

            <div className="card-grid card">
                    {images.map( img => 
                        <GifGridItem 
                            key={img.id} 
                            { ...img }
                        />
                    )}
            </div> 
        </>    
    )
}