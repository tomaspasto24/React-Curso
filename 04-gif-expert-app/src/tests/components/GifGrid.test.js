import { shallow } from "enzyme";
import { GifGrid } from "../../Components/GifGrid";
import useFetchGifs from "../../hooks/useFetchGifs";
import '@testing-library/jest-dom';
jest.mock("../../hooks/useFetchGifs");


describe('Pruebas en GifGrid.js', () => {
    const category = 'One Punch';
    
    test('Debe mostrar el componente correctamente.', () => {

        //Hace que la función que se quiere simular devuelva ese objeto.
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={ category } />);    
        expect(wrapper).toMatchSnapshot();
    });

    test('Se deben mostrar items cuando se cargan imágenes en useFetchGifs', () => {

        const gifs = [{
            id: 'ABC',
            url: 'https://facebook.com',
            title: 'Facebook',
        },
        {
            id: '123',
            url: 'https://google.com',
            title: 'Google',
        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid category={ category } />);    

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBe(false);
        expect( wrapper.find('GifGridItem').length ).toBe( gifs.length );
        
    });

});