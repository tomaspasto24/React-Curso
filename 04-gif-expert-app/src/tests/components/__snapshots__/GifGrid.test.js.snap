// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifGrid.js Debe mostrar el componente correctamente. 1`] = `
<Fragment>
  <h3
    className="card"
  >
     
    One Punch
     
  </h3>
  <p>
    Loading...
  </p>
  <div
    className="card-grid card"
  />
</Fragment>
`;

exports[`Pruebas en GifGrid.js Se deben mostrar items cuando se cargan im√°genes en useFetchGifs 1`] = `
<Fragment>
  <h3
    className="card"
  >
     
    One Punch
     
  </h3>
  <div
    className="card-grid card"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Facebook"
      url="https://facebook.com"
    />
    <GifGridItem
      id="123"
      key="123"
      title="Google"
      url="https://google.com"
    />
  </div>
</Fragment>
`;
